### PR TITLE
kubelet: mark the deprecated pod-infra-container-image flag for removal in 1.35

### PR DESCRIPTION
**Pull request by HirazawaUi**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Prior to version 1.29, there is no workaround for the GC problem, this means the flag served a functionality that falls under the GA deprecation policy.

Based on what [Deprecating a flag or CLI](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli), we seem should deprecate it after 3 releases.  (12months = 3 releases), 1.29 + 3 = 1.32.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by HirazawaUi** [comment]
The test failure seems to be related to directly removing pod-infra-container-image. I think we need to wait for the results of the discussion at https://github.com/kubernetes/kubernetes/pull/122734#discussion_r1450693278 to decide whether to keep pod-infra-container-image args.

**Comment by neolit123** [comment]
```
Rule #5a: CLI elements of user-facing components (e.g. kubectl) must function after their announced deprecation for no less than:

GA: 12 months or 2 releases (whichever is longer)
```

the new CRI functionality was added in 1.29:
https://github.com/kubernetes/kubernetes/pull/118544/commits/d5690f12b69aac5ab4eb0e226682ef68b4059d6e

so the flag can be removed in 1.29 + (12months = 3 releases), 1.29 + 3 = 1.32.

@HirazawaUi i think instead of keeping this PR open in it's current state  i think you can modify it to update the note on top the flag. we can also close the kubeadm PR.

you can then log a ticket in the issue tracker here to update kubelet in 1.32.
for kubeadm the situation is more complicated...
kubeadm cannot be updated anytime soon because kubeadm supports 3 versions of the kubelet (e.g. 1.32,1.31,1.30)
ideally kubeadm should be modified in 1.34 so that t supports kubelet 1.34,1.33,1.32, where 1.32 is the lowest version that doesn't need the flag.

File: pkg/kubelet/config/flags.go [hunk 1]
```diff
@@ -50,8 +46,6 @@ type ContainerRuntimeOptions struct {
 func (s *ContainerRuntimeOptions) AddFlags(fs *pflag.FlagSet) {
 	// General settings.
 	fs.StringVar(&s.RuntimeCgroups, "runtime-cgroups", s.RuntimeCgroups, "Optional absolute name of cgroups to create and run the runtime in.")
-	fs.StringVar(&s.PodSandboxImage, "pod-infra-container-image", s.PodSandboxImage, fmt.Sprintf("Specified image will not be pruned by the image garbage collector. CRI implementations have their own configuration to set this image."))
-	fs.MarkDeprecated("pod-infra-container-image", "will be removed in a future release. Image garbage collector will get sandbox image information from CRI.")
```
[the remark this answers is not in the document]
Comment (HirazawaUi) [reply pkg/kubelet/config/flags.go:54]: okay

**Comment by HirazawaUi** [comment]
> ```
> Rule #5a: CLI elements of user-facing components (e.g. kubectl) must function after their announced deprecation for no less than:
> 
> GA: 12 months or 2 releases (whichever is longer)
> ```
> 
> the new CRI functionality was added in 1.29: [d5690f1](https://github.com/kubernetes/kubernetes/commit/d5690f12b69aac5ab4eb0e226682ef68b4059d6e)
> 
> so the flag can be removed in 1.29 + (12months = 3 releases), 1.29 + 3 = 1.32.
> 
> @HirazawaUi i think instead of keeping this PR open in it's current state i think you can modify it to update the note on top the flag. we can also close the kubeadm PR.
> 
> you can then log a ticket in the issue tracker here to update kubelet in 1.32. for kubeadm the situation is more complicated... kubeadm cannot be updated anytime soon because kubeadm supports 3 versions of the kubelet (e.g. 1.32,1.31,1.30) ideally kubeadm should be modified in 1.34 so that t supports kubelet 1.34,1.33,1.32, where 1.32 is the lowest version that doesn't need the flag.

thanks for reminder.

**Comment by pacoxu** [comment]
It's v1.32 release cycle now. 
- Is this still target v1.32? Or just remove the deprecated flag this release.

**Comment by neolit123** [comment]
reading https://github.com/kubernetes/kubernetes/pull/122739#issuecomment-1969291769 i don't think i got this part correctly or at least i don't remember why i made the comment like that 🤔 

> for kubeadm the situation is more complicated...
kubeadm cannot be updated anytime soon because kubeadm supports 3 versions of the kubelet (e.g. 1.32,1.31,1.30)
ideally kubeadm should be modified in 1.34 so that t supports kubelet 1.34,1.33,1.32, where 1.32 is the lowest version that doesn't need the flag.

potential plan:
- remove the flag in 1.32 from the kubelet
- in kubeadm init/join if the deployed kubelet version is >= 1.32 don't include the flag [1]
- if upgrading using kubeadm upgrade version 1.32, edit the file /var/lib/kubelet/kubeadm-flags.env file and remove the flag
- once kubeadm supports kubelet 1.32 as the oldest version (in 1.32 + 2 = 1.34) we can remove the logic added in [1]

**Comment by SergeyKanzhelev** [comment]
One problem with this is that containerd will be fixing bug with it in 2.0: https://github.com/containerd/containerd/pull/9732

There are no containerd versions that are released that do not have this bug.

There are definitely workaround to pin the preloaded sandbox image directly. But users may not be doing it. Thus removing it from kubelet may break some customers who didn't know about this bug in containerd.

I don't think we are in rush removing arguments, it is not a big deal to have it around. So I would suggest to delay the removal for couple more releases.

**Comment by neolit123** [comment]
> One problem with this is that containerd will be fixing bug with it in 2.0: [containerd/containerd#9732](https://github.com/containerd/containerd/pull/9732)

ok, this is something i did not know about.
thanks for bringing it up.

> I don't think we are in rush removing arguments, it is not a big deal to have it around. So I would suggest to delay the removal for couple more releases.

i think k8s users would need more than 2 releases to switch to 2.0 containerd. from what i've seen containerd upgrades trail back behind k8s upgrades.

**Comment by SergeyKanzhelev** [comment]
@HirazawaUi  let's change the text to `1.35`. Three releases might be fine and we can reevaluate when time will come.

**Comment by HirazawaUi** [comment]
> @HirazawaUi let's change the text to `1.35`. Three releases might be fine and we can reevaluate when time will come.

changed.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Review by neolit123 — COMMENTED** [review]
/hold cancel
LGTM

**Comment by neolit123** [comment]
/retitle kubelet: mark the deprecated pod-infra-container-image flag for removal in 1.35

**Pull request merged**